codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -663,7 +663,7 @@ private void writeHttpBindingMember(
 
             switch (location) {
                 case HEADER:
-                    writer.write("locationName := $S", locationName);
+                    writer.write("locationName := $S", getCanonicalHeader(locationName));
```

Comment:
The reason for this is that using the `Add`/`Get`/`Set` `http.Header` methods always causes a string allocation to occur in order to canonicalize the header. The optimization here is that we pre-compute these during code generation, and have the http encoder utility manipulate the `http.Header` map directly and bypass the `Add` and `Set` appropriately. Users would still be able to use `Get`/`Add`/`Set` in their own custom middleware if they needed to modify ones of these headers as it would correctly match the expect canonicalized form.